### ISSUE TITLE
pretty-printer doesn't consider local terms for unique suffix

### ISSUE THREAD

**Issue by ceedubs** [unisonweb/unison#2699]
The raw transcript is [here](https://gist.github.com/ceedubs/14e1e02c2d645e417f7d077813c0c2fe).

# pretty-printer doesn't consider local terms for unique suffix

It seems that the pretty-printer doesn't consider function-local terms when determining the minimal suffix to reference a term that is declared outside of the function.

Below we add `Foo.bar` and the local `bar`. The watch statement shows that the result is `3`, as would be expected.

```unison
Foo.bar : Nat
Foo.bar = 1

baz : Nat
baz =
  bar = 2
  Foo.bar + bar

> baz
```

```ucm

  I found and typechecked these definitions in scratch.u. If you
  do an `add` or `update`, here's how your codebase would
  change:
  
    ⍟ These new definitions are ok to `add`:
    
      Foo.bar : Nat
      baz     : Nat
  
  Now evaluating any watch expressions (lines starting with
  `>`)... Ctrl+C cancels.

    9 | > baz
          ⧩
          3

```
However, if you use `view` (or `edit`), the pretty-printer shows `bar + bar`, dropping the `Foo` prefix from the first operand.

```ucm
.> add

  ⍟ I've added these definitions:
  
    Foo.bar : Nat
    baz     : Nat

.> view baz

  baz : Nat
  baz =
    use Nat +
    bar = 2
    bar + bar

```

**Comment by atacratic** [comment]
Good bug.  I think this ought to be caught by the `uniqueness` filter in [`calcImports`](https://github.com/unisonweb/unison/blob/1a9b9f80871303bdf9def70c6d23c9d09ba8058c/parser-typechecker/src/Unison/TermPrinter.hs#L1031), but it's not being, because local bindings aren't being included in the `PrintAnnotation.usages` tree of stats about names.  I think the fix would be to add a `LamsNamedOrDelay'` branch to [`suffixCounterTerm`](https://github.com/unisonweb/unison/blob/1a9b9f80871303bdf9def70c6d23c9d09ba8058c/parser-typechecker/src/Unison/TermPrinter.hs#L924) (like the one in [`prettyBinding0`](https://github.com/unisonweb/unison/blob/1a9b9f80871303bdf9def70c6d23c9d09ba8058c/parser-typechecker/src/Unison/TermPrinter.hs#L660) which actually does the pretty-printing of local bindings).  And add tests in the FQN elision zone of the [test suite](https://github.com/unisonweb/unison/blob/1a9b9f80871303bdf9def70c6d23c9d09ba8058c/parser-typechecker/tests/Unison/Test/TermPrinter.hs#L429).  Note also the long explanatory comment [here](https://github.com/unisonweb/unison/blob/1a9b9f80871303bdf9def70c6d23c9d09ba8058c/parser-typechecker/src/Unison/TermPrinter.hs#L784).

Am unlikely to get to this myself ☹️

**Comment by aryairani** [comment]
@ChrisPenner Is there an easy fix in the pretty-printer or in PPE construction for this?

**Comment by ChrisPenner** [comment]
@aryairani There's no simple one-line change AFAICT, we'd need to keep track of all the in-scope locals while pretty-printing in order to do it, then would probably be easiest to just add an argument to the pretty printer of local vars.

Definitely fix-able, but not a super quick win.

Let's look at this in the next triage meeting maybe, if it's high enough priority I could take a look at it 😄

**Comment by ceedubs** [comment]
I think that this was fixed by #4213
